Deprecate version and version_info

facelessuser opened the pull request:
This issue fixes #739. This essentially implements the closest we can get to PEP562 which allows for modules to control `__dir__` and `__getattr__` in order to deprecate attributes. Here we provide a wrapper class for the module in `util`. If a module has attributes that need to deprecated, we derive from the wrapper class and define the attributes as functions with the `property` decorator and the provided `deprecated` decorator. The class is instantiated with the module's `__name__` attribute and the class will properly replace the module with the wrapped module.  When accessing the depracted attributes, a warning will raised.

facelessuser committed:
Deprecate version and version_info

facelessuser commented:
Turns out this works fine with `__all__` even if those items are deprecated in `__all__`. We don't do that, but it is nice to know you get the warning on `from module import *`. 

Let me know if we see improvements that can be made.

waylan commented on markdown/util.py:
```diff
@@ -177,6 +177,39 @@ def code_escape(text):
+        """
+        Implement the `dir` command.
+
+        Return module's `dir` and any attributes that are not private.
```
Not sure we should include `dir`. If we don't, then the deprecated vars won't get listed in the result. But, given they are deprecated, why would we want to make them discoverable?

On the other hand, including `dir` here provides feature parity with Python 3.7 behavior. Guess I'm indifferent about it. I wouldn't have added it, but now that it's here, I'm okay leaving it.

facelessuser replied:
I think not having prevents it from working properly when the items are listed in "all", but we don't have ours there, so it doesn't matter. I can exclude then from dir if you prefer, but I'm also indifferent.

facelessuser commented:
Guess I need a test to cover dir as well.

waylan commented:
And now that it is documented, to address a pet peeve of mine... 😉 ... the [Contributing Guide][1] states:

> If you are providing a fix for a previously reported issue, you must reference the issue in your commit message. Be sure to prefix the reference with one of GitHub's [action words] which will automatically close the issue when the pull request is merged. For example, `fixes #42` and `closes #42` would be acceptable, whereas `ref #42` would not. Of course, if merging a pull request should not cause an issue to be closed, then the action word should not be included when referencing that issue.

[1]: https://github.com/Python-Markdown/markdown/pull/732/files#diff-05b7a8e562e0a482b29460c121776963R78
[action words]: https://help.github.com/articles/closing-issues-using-keywords/

facelessuser commented:
Guess I should read the guide again :smile:.

facelessuser committed:
Test __dir__ attribute of class wrapper

Swap back to refrencing class attributes first to properly expose all attributes.

facelessuser committed:
Python2 and Python3 compatilble __dir__

Python2 cannot access __dir__ with super, and the attribute keys are stored in the base class, not the instance. So to generate a consistent __dir__ in Python 2 and Python 3, we need to get the keys from the derived class base __dict__, and then get the difference from the original class __dict__.

facelessuser commented:
All tests are passing.  I got hung up on Python 2 and Python 3 differences when implementing `__dir__`, but the results are now consistent across all Python versions.  I can't wait till EOL of Python 2.  I also updated the issues message properly refrencing the issue and providing a detailed description.

facelessuser committed:
Simplify legacy version retrieval

There is no need to acces __version__ and __version_info__ through indirection.

waylan commented:
Looks good. Just need to update the release notes and this will be good to go.

facelessuser committed:
Update release notes and comments

Update the 3.0 release notes to contain information on the deprcation of legacy version related attributes. Clean up source comments  related to new deprecation.
